packages/uploadthing/src/client.ts

```diff
@@ -160,114 +175,94 @@ export const DANGEROUS__uploadFiles = async <TRouter extends FileRouter>(
     });
   }
 
-  const fileUploadPromises = s3ConnectionRes.map(async (presigned: any) => {
-    const file = opts.files.find((f) => f.name === presigned.name);
+  const fileUploadPromises = s3ConnectionRes.map(async (presigned) => {
+    const file = opts.files.find((f) => f.name === presigned.fileName);
 
     if (!file) {
       console.error("No file found for presigned URL", presigned);
       throw new UploadThingError({
         code: "NOT_FOUND",
         message: "No file found for presigned URL",
         cause: `Expected file with name ${
-          presigned.name
+          presigned.fileName
         } but got '${opts.files.join(",")}'`,
       });
     }
-    const { url, fields } = presigned.presignedUrl;
-    const formData = new FormData();
-
-    // Give content type to blobs because S3 is dumb
-    // check if content-type is one of the allowed types, or if not and blobs are allowed, use application/octet-stream
-    if (
-      presigned.fileType === file.type.split("/")[0] ||
-      presigned.fileType === file.type
-    ) {
-      formData.append("Content-Type", file.type);
-    } else if (presigned.fileType === "blob") {
-      formData.append("Content-Type", "application/octet-stream");
-    } else if (presigned.fileType === "pdf") {
-      formData.append("Content-Type", "application/pdf");
-    }
-
-    // Dump all values from response (+ the file itself) into form for S3 upload
-    Object.entries({ ...fields, file: file }).forEach(([key, value]) => {
-      formData.append(key, value as Blob);
-    });
 
-    // Do S3 upload
-    const upload = await fetchWithProgress(
-      url,
-      {
-        method: "POST",
-        body: formData,
-        headers: new Headers({
-          Accept: "application/xml",
-        }),
-      },
-      (progressEvent) =>
-        opts.onUploadProgress?.({
-          file: file.name,
-          progress: (progressEvent.loaded / progressEvent.total) * 100,
-        }),
-      () => {
-        opts.onUploadBegin?.({
-          file: file.name,
+    const {
+      presignedUrls,
+      uploadId,
+      chunkSize,
+      contentDisposition,
+      key,
+      pollingUrl,
+    } = presigned;
+
+    const etags = await Promise.all(
+      presignedUrls.map(async (url, index) => {
+        const offset = chunkSize * index;
+        const end = Math.min(offset + chunkSize, file.size);
+        const chunk = file.slice(offset, end);
+
+        const etag = await uploadPartWithProgress({
+          url,
+          chunk: chunk,
+          contentDisposition,
+          fileType: file.type,
+          fileName: file.name,
+          maxRetries: 10,
+          onProgress: (progress) => {
+            const totalUploaded = offset + progress;
+            const percent = (totalUploaded / file.size) * 100;
+            opts.onUploadProgress?.({ file: file.name, progress: percent });
```

Comment:
Should add a check in the components to make sure we always set the progress to a higher value than it was before, otherwise the progress bar jumps around with out-of-order upload of chunks

correct way would probably be to pass the amount to add rather than the absolute percentage since I don't think there is a way for us to know the absolute amount here. 

https://github.com/pingdotgg/uploadthing/blob/dfbe1a85f777648283282b57f303e3ee57f6a46e/packages/react/src/components/button.tsx#L102C1-L102C1